janusgraph-core/src/main/java/org/janusgraph/graphdb/transaction/StandardJanusGraphTx.java

```diff
@@ -1291,8 +1292,12 @@ else if (query.getResultType() == ElementCategory.VERTEX) {
                         }
                     });
                 }
+                //we construct an iterator which lazily streams results from 1st index, and filters by looking up in the intersection of results from all other indices (if any)
+                //NOTE we need to pass NO_LIMIT to processIntersectingRetrievals, otherwise the intersection will be cut down to limit and potentially miss correct results
+                //so this only makes sense if 1st index is the one returning most results (and supporting streaming) AND the rest of indexes are
+                //selective enough to fit into memory and produce full intersection
                 iterator = new SubqueryIterator(indexQuery.getQuery(0), indexSerializer, txHandle, indexCache, indexQuery.getLimit(), getConversionFunction(query.getResultType()),
-                        retrievals.isEmpty() ? null: QueryUtil.processIntersectingRetrievals(retrievals, indexQuery.getLimit()));
+                        retrievals.isEmpty() ? null: QueryUtil.processIntersectingRetrievals(retrievals, Query.NO_LIMIT));
```

Comment:
note that passing NO_LIMIT here is not only required, but also doesn't introduce any additional memory pressure compared to the old implementation (where all indices were passed into it), because internally it makes sure that it "exhausts" all indices i.e. loads all results from them. it just does it in exponential increments based on passed limit, so we actually will save it potentially a lot of effort and memory churn by passing NO_LIMIT upfront